pyopencl/array.py

```diff
@@ -705,6 +736,63 @@ def __init__(
                          "than expected, potentially leading to crashes.",
                          InconsistentOpenCLQueueWarning, stacklevel=2)
 
+    # {{{ Pickling
+
+    def __getstate__(self):
+        try:
+            queue = _QUEUE_FOR_PICKLING_TLS.queue
+        except AttributeError:
+            queue = None
+
+        if queue is None:
+            raise RuntimeError("CL Array instances can only be pickled while "
+                               "queue_for_pickling is active.")
+
+        d = {}
+        d["shape"] = self.shape
+        d["dtype"] = self.dtype
+        d["strides"] = self.strides
+        d["allocator"] = self.allocator
+        d["nbytes"] = self.nbytes
+        d["strides"] = self.strides
+        d["offset"] = self.offset
+        d["data"] = self.get(queue=queue)
+        d["_flags"] = self._flags
+        d["size"] = self.size
+
+        return d
+
+    def __setstate__(self, state):
+        try:
+            queue = _QUEUE_FOR_PICKLING_TLS.queue
+        except AttributeError:
+            queue = None
+
+        if queue is None:
+            raise RuntimeError("CL Array instances can only be pickled while "
+                               "queue_for_pickling is active.")
+
+        self.queue = queue
+        self.shape = state["shape"]
+        self.dtype = state["dtype"]
+        self.strides = state["strides"]
+        self.allocator = state["allocator"]
+        self.offset = state["offset"]
+        self._flags = state["_flags"]
+        self.size = state["size"]
+        self.nbytes = state["nbytes"]
+        self.events = []
+
+        if self.allocator is None:
+            context = queue.context
+            self.base_data = cl.Buffer(context, cl.mem_flags.READ_WRITE, self.nbytes)
+        else:
+            self.base_data = self.allocator(self.nbytes)
```

Comment:
We probably want to allow the user to supply an allocator via the context manager.